Module D: trait objects and Rust patterns (#46)
tweedegolf/101-rs: slides/D-trait-objects-patterns.md

---
theme: default
class: text-center
highlighter: shiki
lineNumbers: true
info: "Rust - D: Trait objects and Rust patterns"
drawings:
  persist: false
fonts:
  mono: Fira Mono
layout: cover
title: 'Rust - D: Trait objects and Rust patterns'
---
# Rust programming
Module D: Trait objects and Rust patterns
<!-- Start with welcome, students entering -->

---
layout: cover
---
# In this module

- Introducing dynamic dispatch
- Work with design pattern in Rust
- Common anti-patterns to avoid

---
layout: default
---
# Learning objectives
<!-- List this module's learning objectives -->

> - Understand the difference between static and dynamic dispatch

Reviewer: I think this is a great opening. All "high concepts".

- Be able to create trait objects
- Understand the concept of 'object safety'
- Apply various commonly used design patterns
- Know which anti-patterns to avoid

---
layout: cover
---
#  Module D
Idiomatic Rust Patterns
<!-- Start lecture content here -->

---
layout: default
---
# Content overview
- Trait objects and dynamic dispatch
- Rust design patterns
- The deref polymorphism anti-pattern
- Project intro

---
layout: section
---
# Trait objects & dynamic dispatch

---
layout: default
---

# Trait... Object?
- We learned about traits in module A3
- We learned about generics and `monomorphization`

> 

Reviewer: I hope you will explain what you mean with that "monomorphization". I had to Google what it means and then went "oh *that*". But perhaps put that a keyword or illustration of it (e.g. a very silly polymorphic function being called on a i32) too.

```rust
fn main() {
    let mut shapes = Vec::new();
    let circle = Circle;
    shapes.push(circle);
    let rect = Rectangle;
    shapes.push(rect);
    shapes.iter().for_each(|shape| shape.paint());
}
```
<v-click>
Becomes

```rust{all|2,3,5}
fn main() {
    let mut shapes: Vec<Box<dyn Render>> = Vec::new();
    let circle = Box::new(Circle);
    shapes.push(circle);
    let rect = Box::new(Rectangle);
    shapes.push(rect);
    shapes.iter().for_each(|shape| shape.paint());
}
```

All set!
</v-click>

---
layout: default
---

# Trait object limitations

- Pointer indirection cost
- Harder to debug
- Type erasure
- Not all traits work:

*Traits need to be 'Object Safe'*

---
layout: default
---

# Object safety

In order for a trait to be object safe, these conditions need to be met:

- If `trait T: Y`, then`Y` must be object safe
- trait `T` must not be `Sized`
- No associated constants allowed
- No associated types with generic allowed
- All associated functions must either be dispatchable from a trait object, or explicitly non-dispatchable

Details in [The Rust Reference](https://doc.rust-lang.org/reference/items/traits.html#object-safety). Read them!

> 

Reviewer: A simple question: *"WHY?"* Now you probably don't want to talk about that (it would probably be 45 minutes lecture in its own right), but the link provided *also* doesn't tell me the "why". Is there a RFC or a book that students can read?

Author: I'll add some content on the 'why'

    }
}
```

---
layout: statement
---

# 1. The Newtype pattern
a small but useful pattern

---
layout: default
---

# Newtype: introduction
&nbsp;

### Wrap an external type in a new local type

```rust
pub struct Imei(String)
```

That's it!

---
layout: default
---

# Newtype: example

```rust
pub enum ValidateImeiError { /* - snip - */}

pub struct Imei(String);

impl Imei {
    fn validate(imei: &str) -> Result<(), ValidateImeiError> {
        todo!();
    }
}

impl TryFrom<String> for Imei {
    type Error = ValidateImeiError;

    fn try_from(imei: String) -> Result<Self, Self::Error> {
        Self::validate(&imei)?;
        Ok(Self(imei))
    }
}

fn register_phone(imei: Imei, label: String) {
    // We can certain `imei` is valid here
}
```

---
layout: default
---

# Newtype: when to use

Newtype solves some problems:
- Orphan rule: no `impl`s for external `trait`s on external types
- Allow for semantic typing (`url` example from mod B)
- Enforce input validation

---
layout: statement
---

# 2. The RAII guard pattern
More robust resource handling

---
layout: default
---

# RAII Guards: introduction

- Resource Acquisition Is Initialization (?)

> - Couple resource aquisition and forfeiture with lifetime of a variable

Reviewer: This is a pretty dense due to "aquistion and forfeiture"
```suggestion
- Link acquiring/releasing a resource to the lifetime of a variable
```

- Access resource through the guard

*Do you know of an example?*

---
layout: two-cols
---

# RAII Guards: example

```rust
pub struct Transaction<'c> {
    connection: &'c mut Connection,
    did_commit: bool,
    id: usize,
}

impl<'c> Transaction<'c> {
    pub fn begin(connection: &'c mut Connection)
     -> Self {
        let id = 
            connection.start_transaction();
        Self {
            did_commit: false,
            id,
            connection,
        }
    }

    pub fn query(&self sql: &str) { /* - snip - */}

    pub fn commit(self) {
        self.did_commit = true;
    }
}
```
::right::
<div style="padding-left:10px; padding-top: 50px;">

```rust
impl Drop for Transaction<'_> {
    fn drop(&mut self) {
        if self.did_commit {
            self
                .connection
                .commit_transaction(self.id);
            
        } else {
            self
                .connection
                .rollback_transaction(self.id);
        }
    }
}
```
</div>

---
layout: default
---

# RAII Guards: when to use

- Ensure a resource is freed at some point
- Ensure invariants hold while guard lives

---
layout: statement
---

> # 3. The Typestate pattern

Reviewer: When you get to this point, take stock of how much time you have left (and want to spend on things after this section). The typestate pattern is nice, but in principle people can also explore this in their own time (or during the project).

Reviewer: I.e. you don't want to rush the "anti-patterns" because you lost the time in this section.

struct Animal {
    name: String,
}

impl Animal {
    fn walk(&self) {
        println!("Tippy tap")
    }
    fn eat(&self) {
        println!("Om nom")
    }
    fn say_name(&self) {
        // Animals generally can't speak
        println!("...")
    }
}
```
::right::
```rust
struct Dog {
    animal: Animal
}
impl Dog {
    fn eat(&self) {
        println!("Munch munch");
    }
    fn bark(&self) {
        println!("Woof woof!");
    }
}

impl Deref for Dog {
    type Target = Animal;

    fn deref(&self) -> &Self::Target {
        &self.animal
    }
}

fn main (){ 
    let dog: Dog = todo!("Instantiate Dog");
    dog.bark();
    dog.walk();
    dog.eat();
    dog.say_name();
}
```

---
layout: default
---

# The output

```txt
Woof woof!
Tippy tap
Munch munch
```

*Even overloading works!*

---
layout: default
---

# Why is it bad?

- This is no 'real' inheritance: `Dog` is no subtype of `Animal`
- Traits implemented on `Animal` are not implemented on `Dog` automatically
- `Deref` and `DerefMut` are intended 'pointer-to-`T`' to `T` conversions
- Deref coercion by `.` 'converts' `self` from `Dog` to `Animal`
- Rust favours explicit conversions for easier reasoning about code

> *It will only add confusion: for OOP programmers it's incomplete, for Rust programmers it goes against semantics*

Reviewer: In my ears, if something goes against the *semantics*, that means "this doesn't mean what you think it means, it prints 5 instead of 2". Here the code shown is a different beast; it's more that it goes against the "pragmatics": yes you can do this, but it will create more problems than it solves. Suggestion:

```suggestion
*It will only add confusion: for OOP programmers it's incomplete, for Rust programmers it is unidiomatic*
```
